File: examples/streaming_client.py
"""
Example client for Sutra AI streaming API.

Demonstrates:
- Progressive answer refinement
- Real-time confidence updates
- Handling of streaming stages
"""

import asyncio
import json
import httpx


async def stream_query_simple(query: str, api_url: str = "http://localhost:8001"):
    """
    Simple streaming client - prints results as they arrive.
    
    Args:
        query: Question to ask
        api_url: Base URL of Sutra Hybrid API
    """
    url = f"{api_url}/sutra/stream/query"
    
    async with httpx.AsyncClient(timeout=30.0) as client:
        async with client.stream(
            "POST",
            url,
            json={"query": query, "max_concepts": 10}
        ) as response:
            print(f"\n🤔 Query: {query}")
            print("=" * 60)
            
            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    data = line[6:]  # Remove "data: " prefix
                    try:
                        chunk = json.loads(data)
                        
                        # Print progress
                        stage = chunk.get('stage', 'unknown')
                        answer = chunk.get('answer', '')
                        confidence = chunk.get('confidence', 0.0)
                        paths = chunk.get('paths_found', 0)
                        is_final = chunk.get('is_final', False)
                        
                        # Stage indicator
                        if stage == 'initial':
                            print(f"\n⚡ Initial answer (1 path):")
                        elif stage == 'refining':
                            print(f"\n🔄 Refining... ({paths} paths):")
                        elif stage == 'consensus':
                            print(f"\n🎯 Consensus ({paths} paths):")
                        elif stage == 'complete':
                            print(f"\n✅ Final answer ({paths} paths):")
                        
                        # Answer
                        print(f"   {answer}")
                        print(f"   Confidence: {confidence:.2f}")
                        
                        if is_final:
                            print("\n" + "=" * 60)
                            break
                    
                    except json.JSONDecodeError:
                        pass


async def stream_query_with_callback(
    query: str,
    on_chunk,
    api_url: str = "http://localhost:8001"
):
    """
    Streaming client with callback for custom handling.
    
    Args:
        query: Question to ask
        on_chunk: Callback function called for each chunk
        api_url: Base URL of Sutra Hybrid API
    """
    url = f"{api_url}/sutra/stream/query"
    
    async with httpx.AsyncClient(timeout=30.0) as client:
        async with client.stream(
            "POST",
            url,
            json={"query": query, "max_concepts": 10}
        ) as response:
            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    data = line[6:]
                    try:
                        chunk = json.loads(data)
                        await on_chunk(chunk)
                        
                        if chunk.get('is_final'):
                            break
                    except json.JSONDecodeError:
                        pass


async def stream_query_collect_all(
    query: str,
    api_url: str = "http://localhost:8001"
) -> list:
    """
    Collect all streaming chunks into a list.
    
    Useful for analysis or replay.
    
    Args:
        query: Question to ask
        api_url: Base URL of Sutra Hybrid API
    
    Returns:
        List of all chunks received
    """
    chunks = []
    
    async def collect(chunk):
        chunks.append(chunk)
    
    await stream_query_with_callback(query, collect, api_url)
    return chunks


# Example usage
async def main():
    """Run example streaming queries."""
    
    # Example 1: Simple streaming
    print("\n" + "=" * 60)
    print("Example 1: Simple Streaming")
    print("=" * 60)
    await stream_query_simple("What is machine learning?")
    
    # Example 2: With custom callback
    print("\n" + "=" * 60)
    print("Example 2: Custom Callback")
    print("=" * 60)
    
    async def my_callback(chunk):
        stage = chunk['stage']
        conf = chunk['confidence']
        print(f"[{stage.upper()}] Confidence: {conf:.2f}")
    
    await stream_query_with_callback(
        "What is artificial intelligence?",
        my_callback
    )
    
    # Example 3: Collect all chunks
    print("\n" + "=" * 60)
    print("Example 3: Collect All Chunks")
    print("=" * 60)
    
    chunks = await stream_query_collect_all("What is deep learning?")
    print(f"Received {len(chunks)} chunks")
    print(f"Confidence progression: {[c['confidence'] for c in chunks]}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/streaming_client.js
/**
 * JavaScript/TypeScript example for Sutra AI streaming API.
 * 
 * Works in:
 * - Modern browsers
 * - Node.js (with fetch polyfill)
 * - React/Vue/Angular applications
 */

/**
 * Stream query with progressive answer refinement.
 * 
 * @param {string} query - Question to ask
 * @param {object} options - Streaming options
 * @param {function} options.onChunk - Called for each chunk
 * @param {function} options.onComplete - Called when stream completes
 * @param {function} options.onError - Called on error
 * @param {string} options.apiUrl - Base API URL
 */
async function streamQuery(query, options = {}) {
    const {
        onChunk = (chunk) => console.log(chunk),
        onComplete = () => console.log('Stream complete'),
        onError = (err) => console.error('Stream error:', err),
        apiUrl = 'http://localhost:8001'
    } = options;

    const url = `${apiUrl}/sutra/stream/query`;

    try {
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                query: query,
                max_concepts: 10,
                enable_quality_gates: true
            })
        });

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();

        let buffer = '';

        while (true) {
            const { done, value } = await reader.read();
            
            if (done) {
                break;
            }

            // Decode chunk and add to buffer
            buffer += decoder.decode(value, { stream: true });

            // Process complete messages
            const lines = buffer.split('\n');
            buffer = lines.pop(); // Keep incomplete line in buffer

            for (const line of lines) {
                if (line.startsWith('data: ')) {
                    const data = line.slice(6);
                    try {
                        const chunk = JSON.parse(data);
                        onChunk(chunk);

                        if (chunk.is_final) {
                            onComplete(chunk);
                            return;
                        }
                    } catch (e) {
                        // Ignore parse errors for non-JSON data
                    }
                } else if (line.startsWith('event: error')) {
                    const errorData = lines[lines.indexOf(line) + 1];
                    if (errorData && errorData.startsWith('data: ')) {
                        const error = JSON.parse(errorData.slice(6));
                        onError(error);
                        return;
                    }
                }
            }
        }
    } catch (error) {
        onError(error);
    }
}


/**
 * React hook for streaming queries.
 * 
 * Usage:
 *   const { answer, confidence, isLoading, stream } = useStreamingQuery();
 *   
 *   // In component:
 *   <button onClick={() => stream("What is AI?")}>Ask</button>
 *   {isLoading && <div>Loading... ({confidence})</div>}
 *   <div>{answer}</div>
 */
function useStreamingQuery() {
    const [answer, setAnswer] = React.useState('');
    const [confidence, setConfidence] = React.useState(0);
    const [stage, setStage] = React.useState('');
    const [isLoading, setIsLoading] = React.useState(false);
    const [error, setError] = React.useState(null);

    const stream = async (query, apiUrl = 'http://localhost:8001') => {
        setIsLoading(true);
        setError(null);
        setAnswer('');
        setConfidence(0);

        await streamQuery(query, {
            onChunk: (chunk) => {
                setAnswer(chunk.answer);
                setConfidence(chunk.confidence);
                setStage(chunk.stage);
            },
            onComplete: () => {
                setIsLoading(false);
            },
            onError: (err) => {
                setError(err);
                setIsLoading(false);
            },
            apiUrl
        });
    };

    return { answer, confidence, stage, isLoading, error, stream };
}


/**
 * Vue 3 Composition API example.
 */
function useStreamingQueryVue() {
    const answer = ref('');
    const confidence = ref(0);
    const stage = ref('');
    const isLoading = ref(false);
    const error = ref(null);

    const stream = async (query, apiUrl = 'http://localhost:8001') => {
        isLoading.value = true;
        error.value = null;
        answer.value = '';
        confidence.value = 0;

        await streamQuery(query, {
            onChunk: (chunk) => {
                answer.value = chunk.answer;
                confidence.value = chunk.confidence;
                stage.value = chunk.stage;
            },
            onComplete: () => {
                isLoading.value = false;
            },
            onError: (err) => {
                error.value = err;
                isLoading.value = false;
            },
            apiUrl
        });
    };

    return { answer, confidence, stage, isLoading, error, stream };
}


/**
 * Vanilla JavaScript example with UI updates.
 */
function createStreamingUI(containerId) {
    const container = document.getElementById(containerId);
    
    container.innerHTML = `
        <div class="streaming-container">
            <input type="text" id="query-input" placeholder="Ask a question..." />
            <button id="ask-button">Ask</button>
            <div id="status"></div>
            <div id="answer-container">
                <div id="stage"></div>
                <div id="answer"></div>
                <div id="confidence"></div>
                <div id="paths"></div>
            </div>
        </div>
    `;

    const input = document.getElementById('query-input');
    const button = document.getElementById('ask-button');
    const status = document.getElementById('status');
    const stageEl = document.getElementById('stage');
    const answerEl = document.getElementById('answer');
    const confidenceEl = document.getElementById('confidence');
    const pathsEl = document.getElementById('paths');

    button.addEventListener('click', async () => {
        const query = input.value.trim();
        if (!query) return;

        button.disabled = true;
        status.textContent = 'Streaming...';

        await streamQuery(query, {
            onChunk: (chunk) => {
                // Stage indicator with emoji
                const stageEmoji = {
                    'initial': '⚡',
                    'refining': '🔄',
                    'consensus': '🎯',
                    'complete': '✅'
                };

                stageEl.textContent = `${stageEmoji[chunk.stage] || '❓'} ${chunk.stage}`;
                answerEl.textContent = chunk.answer;
                confidenceEl.textContent = `Confidence: ${(chunk.confidence * 100).toFixed(0)}%`;
                pathsEl.textContent = `Paths: ${chunk.paths_found}`;
            },
            onComplete: () => {
                status.textContent = 'Complete!';
                button.disabled = false;
            },
            onError: (err) => {
                status.textContent = `Error: ${err.message}`;
                button.disabled = false;
            }
        });
    });
}


// Node.js example
async function nodeExample() {
    console.log('Streaming query example:\n');

    await streamQuery('What is machine learning?', {
        onChunk: (chunk) => {
            const emoji = {
                'initial': '⚡',
                'refining': '🔄',
                'consensus': '🎯',
                'complete': '✅'
            };

            console.log(`\n${emoji[chunk.stage]} ${chunk.stage.toUpperCase()}`);
            console.log(`Answer: ${chunk.answer}`);
            console.log(`Confidence: ${(chunk.confidence * 100).toFixed(0)}%`);
            console.log(`Paths: ${chunk.paths_found}`);
        },
        onComplete: () => {
            console.log('\n✅ Stream complete!');
        },
        onError: (err) => {
            console.error('\n❌ Error:', err);
        }
    });
}

// Export for use in modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        streamQuery,
        useStreamingQuery,
        useStreamingQueryVue,
        createStreamingUI
    };
}

// Run example if executed directly
if (typeof require !== 'undefined' && require.main === module) {
    nodeExample();
}
